rfcs/xxxx-custom-event-metadata-from-annotations/README.md

```diff
@@ -0,0 +1,184 @@
+# RFC-0007 Custom Event Metadata from Annotations
+
+**Status:** provisional
+
+<!--
+Status represents the current state of the RFC.
+Must be one of `provisional`, `implementable`, `implemented`, `deferred`, `rejected`, `withdrawn`, or `replaced`.
+-->
+
+**Creation date:** 2024-05-23
+
+**Last update:** 2024-05-23
+
+## Summary
+
+Flux users often run into situations where they wish to send custom, static metadata fields defined
+in Flux objects on the events dispatched by the respective controller to notification-controller.
+This proposal offers a solution for supporting those use cases uniformly across all Flux controllers
+by sending annotation keys that are prefixed with a well-defined API Group.
+
+## Motivation
+
+This RFC comes as a response to the need for adding custom metadata to events about Flux objects
+sent to notification providers. See specific user stories in the [User Stories](#user-stories) section.
+
+### Goals
+
+Provide a method for Flux users to embed custom/static metadata in their Flux objects
+and have that metadata propagated to the notification providers.
+
+### Non-Goals
+
+In this proposal we **do not** aim to provide a method for Flux users to send etcd-indexed custom metadata
+fields from Flux objects in events to notification-controller, most specifically labels. By design an event
+already contains enough identification information to locate the associated Flux object inside the cluster,
+which covers the use case of labels. Flux does not wish to incentivize practices that are impactful to clusters
+without a strong reason or benefit.
+
+## Proposal
+
+When sending events about Flux objects, we propose sending annotation keys prefixed with a well-defined
+API Group in addition to those already sent by the controller.
+
+### User Stories
+
+#### Story 1
+
+> As a user, I want to embed FluxCD into my GitHub Workflow in a way that it only succeeds if
+> the deployment made by FluxCD is successful.
+
+For example, embedding a Deployment ID from the GitHub API in a `HelmRelease` object like the one below:
+
+```yaml
+apiVersion: helm.toolkit.fluxcd.io/v2
+kind: HelmRelease
+metadata:
+  name: podinfo
+  namespace: flux-system
+  annotations:
+    event.toolkit.fluxcd.io/deploymentID: e076e315-5a48-41c3-81c8-8d8bdee7d74d
```

Comment:
Not sure this use-case is valid, how would people know the deployment ID as it changes with each PR?

Reply:
The deployment ID in my example is determined by the workflow which triggers the deployment. Basically, it is set during the workflow. Yes, it changes with each PR, but that is the goal in my use case. I want to be able to map a Flux deployment to the workflow which triggered it (caused the state change).

Reply:
> Basically, it is set during the workflow. 

Are you creating the HelmRelease directly on the cluster from GitHub Actions?

Reply:
I love this idea. We have the similar needs where we would to know the origin of a change more easily. https://github.com/fluxcd/flux2/discussions/4718#discussioncomment-10984304